File: src/api/schemas/mcp.py
"""API schemas for MCP (Model Context Protocol) servers."""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from src.mcp.models import ServerStatus


# ========== Catalog Schemas ==========


class CredentialSpecSchema(BaseModel):
    """Schema for credential specification."""

    name: str
    description: str
    sensitive: bool
    header_name: Optional[str] = None


class CatalogTemplateSchema(BaseModel):
    """Schema for MCP server template from catalog."""

    id: str
    name: str
    url_template: Optional[str] = None
    auth_type: str
    token_guide_url: Optional[str] = None
    scopes: List[str] = []
    credentials_required: List[CredentialSpecSchema]
    credentials_optional: List[CredentialSpecSchema] = []
    tools: List[str] = []


class CatalogListResponse(BaseModel):
    """Response containing catalog templates."""

    servers: List[CatalogTemplateSchema]
    total: int


# ========== Server Management Schemas ==========


class CreateServerFromTemplateRequest(BaseModel):
    """Request to create MCP server from catalog template."""

    template: str = Field(..., description="Template ID from catalog")
    name: str = Field(..., min_length=1, max_length=200)
    credentials: Dict[str, str] = Field(
        ...,
        description="Credentials as key-value pairs (e.g., GOOGLE_REFRESH_TOKEN)",
    )
    headers: Optional[Dict[str, str]] = Field(
        default=None,
        description="Additional non-sensitive headers",
    )


class CreateCustomServerRequest(BaseModel):
    """Request to create custom MCP server."""

    name: str = Field(..., min_length=1, max_length=200)
    url: str = Field(..., description="MCP server URL (Streamable HTTP endpoint)")
    credentials: Dict[str, str] = Field(
        default_factory=dict,
        description="Credentials as key-value pairs",
    )
    headers: Optional[Dict[str, str]] = Field(
        default=None,
        description="Additional HTTP headers",
    )


class MCPServerResponse(BaseModel):
    """Response containing MCP server details."""

    id: str
    name: str
    template: Optional[str] = None
    url: str
    status: ServerStatus
    created_at: datetime
    last_used: Optional[datetime] = None
    error_message: Optional[str] = None

    class Config:
        from_attributes = True


class MCPServerListResponse(BaseModel):
    """Response containing list of MCP servers."""

    servers: List[MCPServerResponse]
    total: int


class MCPServerToolSchema(BaseModel):
    """Schema for tool provided by MCP server."""

    name: str
    description: str
    input_schema: Dict[str, Any]


class MCPServerDetailResponse(BaseModel):
    """Detailed response with server info and available tools."""

    id: str
    name: str
    template: Optional[str] = None
    url: str
    status: ServerStatus
    created_at: datetime
    last_used: Optional[datetime] = None
    error_message: Optional[str] = None
    tools: List[MCPServerToolSchema] = []

    class Config:
        from_attributes = True


# ========== Health & Status Schemas ==========


class ServerHealthStatus(BaseModel):
    """Health status for a single server."""

    server_id: str
    name: str
    status: ServerStatus
    error: Optional[str] = None


class MCPHealthResponse(BaseModel):
    """Response containing health status of all MCP servers."""

    servers: List[ServerHealthStatus]
    total_active: int
    total_error: int
    total_disconnected: int

File: web/types/mcp.ts
export type AuthType = "oauth_token" | "api_key" | "none";

export interface CredentialSpec {
  name: string;
  description: string;
  sensitive: boolean;
  header_name?: string;
}

export interface MCPCatalogItem {
  id: string;
  name: string;
  url_template: string;
  auth_type: AuthType;
  token_guide_url?: string;
  scopes?: string[];
  credentials_required: CredentialSpec[];
  credentials_optional: CredentialSpec[];
  tools: string[];
}

export interface MCPCatalogResponse {
  servers: MCPCatalogItem[];
  total: number;
}

export type ServerStatus = "active" | "error" | "disconnected";

export interface MCPServer {
  id: string;
  name: string;
  template?: string;
  url: string;
  status: ServerStatus;
  created_at: string;
  last_used?: string;
  error_message?: string;
}

export interface MCPServerDetail extends MCPServer {
  tools: MCPTool[];
}

export interface MCPTool {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

export interface MCPServerListResponse {
  servers: MCPServer[];
  total: number;
}

export interface CreateMCPServerFromTemplate {
  template: string;
  name: string;
  credentials: Record<string, string>;
  oauth_token_ref?: string;
}

export interface CreateMCPServerCustom {
  name: string;
  url: string;
  credentials?: Record<string, string>;
  headers?: Record<string, string>;
}

export type CreateMCPServerRequest = CreateMCPServerFromTemplate | CreateMCPServerCustom;

export interface MCPHealthStatus {
  server_id: string;
  name: string;
  status: ServerStatus;
  error?: string;
}

export interface MCPHealthResponse {
  servers: MCPHealthStatus[];
  total_active: number;
  total_error: number;
  total_disconnected: number;
}

export interface OAuthAuthorizeResponse {
  authorization_url: string;
  service: string;
}

export interface OAuthCallbackResponse {
  refresh_token: string;
  service: string;
  message: string;
}
